rmf_site_editor/src/site/model.rs

```diff
@@ -16,528 +16,659 @@
+    children_q: Query<&Children>,
+    models: Query<&AssetSource, With<ModelMarker>>,
+    model_loading: Res<ModelLoadingServices>,
+) -> impl Future<Output = Result<ModelLoadingRequest, ModelLoadingError>> {
+    let models = DescendantIter::new(&children_q, request.parent)
+        .filter_map(|c| models.get(c).ok().map(|source| (c, source.clone())))
+        .collect::<Vec<_>>();
+    let load_model = model_loading.load_model.clone();
+    async move {
+        for (model_entity, source) in models {
+            channel
+                .query(ModelLoadingRequest::new(model_entity, source), load_model)
+                .await
+                .available()
+                .ok_or_else(|| {
+                    ModelLoadingError::new(
+                        request.clone(),
+                        ModelLoadingErrorKind::WorkflowExecutionError,
+                    )
+                })?
+                .map_err(|err| {
+                    ModelLoadingError::new(
+                        request.clone(),
+                        ModelLoadingErrorKind::FailedLoadingDependency(err.to_string()),
+                    )
+                })?;
+        }
+        Ok(request)
+    }
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, Clone, SystemSet)]
+pub enum ModelLoadingSet {
+    /// Label to run the system that checks if scenes have been spawned
+    CheckSceneSystem,
+    /// Flush commands and impulses
+    CheckSceneFlush,
+}
 
-impl FromWorld for ModelTrashcan {
-    fn from_world(world: &mut World) -> Self {
-        Self(world.spawn(Trashcan).id())
+impl ModelLoadingServices {
+    pub fn from_app(app: &mut App) -> Self {
+        app.configure_sets(
+            PostUpdate,
+            (
+                ModelLoadingSet::CheckSceneSystem,
+                ModelLoadingSet::CheckSceneFlush,
+            )
+                .chain(),
+        )
+        .add_systems(
+            PostUpdate,
+            (apply_deferred, flush_impulses()).in_set(ModelLoadingSet::CheckSceneFlush),
+        );
+        let check_scene_is_spawned = app.spawn_continuous_service(
+            PostUpdate,
+            check_scenes_are_spawned.configure(|config: SystemConfigs| {
+                config.in_set(ModelLoadingSet::CheckSceneSystem)
+            }),
+        );
+        let skip_if_unchanged = cleanup_if_asset_source_changed.into_blocking_callback();
+        let load_model_dependencies = app.world.spawn_service(load_model_dependencies);
+        let model_loading_service = app.world.spawn_service(handle_model_loading);
+        // This workflow tries to load the model without doing any error handling
+        let try_load_model: Service<ModelLoadingRequest, ModelLoadingResult, ()> =
+            app.world.spawn_workflow(|scope, builder| {
+                scope
+                    .input
+                    .chain(builder)
+                    .then(skip_if_unchanged)
+                    .map_block(|success| {
```

Comment:
I have no problem at all with keeping this as-is, but I thought I'd point out that we could roll this logic into `skip_if_unchanged`. Essentially `cleanup_if_asset_source_changed` could return a `Result<ModelLoadingRequest, ModelLoadingResult>` depending on whether or not a change is needed, and we just do `.branch_for_err(...)` right off of the output of `skip_if_unchanged`.

Reply:
Nice find! https://github.com/open-rmf/rmf_site/pull/238/commits/760e80f0557e541e50f7914dfc753fc27a08e6c9